pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/BaseRecordExtractor.java

```diff
@@ -0,0 +1,190 @@
+   * {@code null}.
+   *
+   * This implementation converts the Collection to an Object array. Override this method if the data format
+   * requires a different conversion for its multi-value objects.
+   *
+   * @param value should be verified to be a Collection type prior to calling this method as it will be casted
+   *              to a Collection without checking
+   */
+  @Nullable
+  protected Object convertMultiValue(Object value) {
+    Collection collection = (Collection) value;
+    if (collection.isEmpty()) {
+      return null;
+    }
+
+    int numValues = collection.size();
+    Object[] array = new Object[numValues];
+    int index = 0;
+    for (Object element : collection) {
+      Object convertedValue = null;
+      if (element != null) {
+        convertedValue = convert(element);
+      }
+      if (convertedValue != null && !convertedValue.toString().equals("")) {
+        array[index++] = convertedValue;
+      }
+    }
+
+    if (index == numValues) {
+      return array;
+    } else if (index == 0) {
+      return null;
+    } else {
+      return Arrays.copyOf(array, index);
+    }
+  }
+
+  /**
+   * Handles the conversion of every value of the map. Note that map keys will be handled as a single-value type.
+   * Returns {@code null} if the field value is {@code null}. This should be overridden if the data format requires
+   * a different conversion for map values.
+   *
+   * @param value should be verified to be a Map type prior to calling this method as it will be casted to a Map
+   *              without checking
+   */
+  @Nullable
+  protected Object convertMap(Object value) {
+    Map<Object, Object> map = (Map) value;
+    if (map.isEmpty()) {
+      return null;
+    }
+
+    Map<Object, Object> convertedMap = new HashMap<>();
+    for (Map.Entry<Object, Object> entry : map.entrySet()) {
+      Object mapKey = entry.getKey();
+      Object mapValue = entry.getValue();
+      if (mapKey != null) {
+        Object convertedMapValue = null;
+        if (mapValue != null) {
+          convertedMapValue = convert(mapValue);
+        }
+
+        if (convertedMapValue != null) {
+          convertedMap.put(convertSingleValue(entry.getKey()), convertedMapValue);
+        }
+      }
+    }
+
+    if (convertedMap.isEmpty()) {
+      return null;
+    }
+
+    return convertedMap;
+  }
+
+  /**
+   * Converts single value types. This should be overridden if the data format requires
+   * a different conversion for its single values.
+   */
+  protected Object convertSingleValue(Object value) {
+    if (value instanceof ByteBuffer) {
+      ByteBuffer byteBufferValue = (ByteBuffer) value;
+
+      // Use byteBufferValue.remaining() instead of byteBufferValue.capacity() so that it still works when buffer is
+      // over-sized
+      byte[] bytesValue = new byte[byteBufferValue.remaining()];
+      byteBufferValue.get(bytesValue);
+      return bytesValue;
+    }
+    if (value instanceof Number) {
```

Comment:
We should also preserve byte[]
```suggestion
    if (value instanceof Number || value instanceof byte[]) {
```

Reply:
Yea that makes sense given that bytebuffer is converted to byte[].